### PR TITLE
Add pre-commit hook to fix lint errors

### PR DESCRIPTION
**Pull request by ethanmills**
### Context
@kicferk1  told me to

### Changes proposed in this pull request
Add a precommit hook which fixes lint errors, and prevents the commit if there are unfixable errors
We will need to run `git config core.hooksPath .githooks` to use these hooks, per the updated readme.

### Guidance to review
Might be worth checking it out locally and seeing is the delay is too long for your liking. Also testing it on a mac would be good....

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+# stash unstaged changes
+git diff > unstaged.diff
+git apply -R unstaged.diff
+
+# Lint, fix, store exit code, and stage modifications
+rubocop --fail-fast --auto-correct --except Layout/EndOfLine
```
Comment (kicferk1) [inline .githooks/pre-commit:8]: We don't care about `Layout/EndOfLine`? 
Is this rubocop using the config in `.rubocop.yml`? If so, can we move it there if we really care? Preferably with a comment why it's needed (windows being annoying I'm guessing).

**Comment by kicferk1** [comment]
> ### Context
> @kicferk1 told me to

Best type of context :D

> ### Guidance to review
> Might be worth checking it out locally and seeing is the delay is too long for your liking. Also testing it on a mac would be good....
Will do tomorrow morning :)

File: .githooks/pre-commit (hunk 1 above)
Comment (ethanmills) [reply .githooks/pre-commit:8]: It's some interaction between windows, rubocop, and git core.autocrlf

**Comment by kicferk1** [comment]
For mac, you also need to run 
```
chmod 755 .githooks/*
```
because 'in appleland files are not executable by default, you naughty user of insecure operating systems'

**Comment by ethanmills** [comment]
> For mac, you also need to run
> 
> ```
> chmod 755 .githooks/*
> ```
> 
> because 'in appleland files are not executable by default, you naughty user of insecure operating systems'

I think git tracks file permissions. Would you be able to commit the 755'd file please?

File: .githooks/pre-commit (hunk 1 above)
Comment (kicferk1) [reply .githooks/pre-commit:8]: Could we try running rubocop only on changed files? 

https://medium.com/devnetwork/running-rubocop-only-on-modified-files-a21aed86e06d has some ugly commands :P

Comment (ethanmills) [reply .githooks/pre-commit:8]: I only want it to run on staged changes, because I often commit parts of a file

Comment (ethanmills) [reply .githooks/pre-commit:8]: This is what lines 4, 5, 18 & 19 do

Comment (ethanmills) [reply .githooks/pre-commit:8]: Oh wait, I see what you mean... Will do

**Review by kicferk1 — APPROVED** [review]
LGTM (and I checked it works on a mac)

**Pull request merged**